abi-versions/vNEXT/README.md

```diff
@@ -0,0 +1,1471 @@
+
+Returns `CompareAndSwapMismatch` when the compare-and-switch value isn't
+current.
+
+Returns `InvalidMemoryAccess` when `key_data` and/or `key_size` point to invalid
+memory address.
+
+#### `proxy_delete_shared_kvstore`
+
+* params:
+  - `i32 (uint32_t) kvstore_id`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Deletes previously created shared key-value store (`kvstore_id`).
+
+Returns `Ok` on success.
+
+Returns `BadArgument` when the key-value store (`kvstore_id`) doesn't exist.
+
+
+## Shared queue
+
+Note: Global queue can be always accessed using `queue_id` `0`.
+
+### Functions exposed by the host environment
+
+#### `proxy_open_shared_queue`
+
+* params:
+  - `i32 (const char*) queue_name_data`
+  - `i32 (size_t) queue_name_size`
+  - `i32 (bool) create_if_not_exist`
+  - `i32 (uint32_t*) return_queue_id`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Opens shared queue with the given name (`queue_name_data`, `queue_name_size`).
+If `create_if_not_exist` is `true` and there is no shared queue with that name,
+then it will be created.
+
+Items can be queued using `proxy_enqueue_shared_queue_item` and dequeued using
+`proxy_dequeue_shared_queue_item` from the queue using returned unique queue
+identifier (`return_queue_id`).
+
+Returns `Ok` on success.
+
+Returns `NotFound` when `create_if_not_exist` is `false` and no shared queue
+with the given name exists.
+
+Returns `InvalidMemoryAccess` when `queue_name_data`, `queue_name_size` and/or
+`return_queue_id` point to invalid memory address.
+
+#### `proxy_dequeue_shared_queue_item`
+
+* params:
+  - `i32 (uint32_t) queue_id`
+  - `i32 (iovec_array_t*) return_payload_data`
+  - `i32 (size_t*) return_payload_size`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Removes item from the tail of the queue (`queue_id`). Returned value is written
+into memory slice (`payload_data`, `payload_size`).
+
+Returns `Ok` on success.
+
+Returns `BadArgument` when the queue (`queue_id`) doesn't exist.
+
+Returns `InvalidMemoryAccess` when `return_payload_data` and/or
+`return_payload_size` point to invalid memory address.
+
+#### `proxy_enqueue_shared_queue_item`
+
+* params:
+  - `i32 (uint32_t) queue_id`
+  - `i32 (iovec_array_t) payload_data`
+  - `i32 (size_t) payload_size`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Adds item (`payload_data`, `payload_size`) at the front of the queue
+(`queue_id`).
+
+Returns `Ok` on success.
+
+Returns `BadArgument` when the queue (`queue_id`) doesn't exist.
+
+Returns `InvalidMemoryAccess` when `payload_data` and/or `payload_size` point to
+invalid memory address.
+
+#### `proxy_delete_shared_queue`
+
+* params:
+  - `i32 (uint32_t) queue_id`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Deletes previously created shared queue (`queue_id`).
+
+Returns `Ok` on success.
+
+Returns `BadArgument` when the queue (`queue_id`) doesn't exist.
+
+### Callbacks optionally exposed by the Wasm module
+
+#### `proxy_on_queue_ready`
```

Comment:
It's not correlated with any `context_id`, since queue is a standalone object.

`proxy_open_shared_queue` returns unique `queue_id` handle, which is used in `proxy_on_queue_ready`.

Always notifying all WasmVms seems wasteful, so we should do some kind of load-balancing (probably with ACK/NACK), but I guess we could declare each queue as either unicast or multicast upon creation, if you have a use case for it?